lib/elixir/lib/kernel.ex

```diff
@@ -3069,6 +3076,36 @@ defmodule Kernel do
     end
   end
 
+  def ensure_evaled(list, i) do
+    :lists.mapfoldl(fn
+      elem, acc when is_integer(elem) or is_atom(elem) or is_binary(elem) ->
+        {elem, acc}
+      {:|, meta, [head, tail]}, {i, ast} when is_integer(head) or is_atom(head) or is_binary(head) ->
+        {tail, {i, tail_ast}} = ensure_evaled(tail, i)
```

Comment:
I just realized though that this works today:

    1 in [0, 2 | 1..3]

which is obviously undesired. We need to fix this by changing `in_list` to compile to something that handles exclusively lists. Possibly `:lists.member/2`.

Here is a version that removes the duplication and avoids changing the tail:

```elixir
    # Called as ensure_evaled(list, {0, []}). Note acc is reversed.
    def ensure_evaled(list, acc) do
      :lists.mapfoldl(fn
        {:|, meta, [head, tail]}, acc ->
          {head, acc} = ensure_evaled_head(head, acc)
          {tail, acc} = ensure_evaled_tail(tail, acc)
          {{:|, meta, [head, tail]}, acc}
        elem, acc ->
          ensure_evaled_head(elem, acc)
      end, acc, list)
    end

    defp ensure_evaled_head(elem, acc) when is_integer(elem) or is_atom(elem) or is_binary(elem) do
      {elem, acc}
    end
    defp ensure_evaled_head(elem, acc) do
      ensure_evaled_var(elem, acc)
    end

    defp ensure_evaled_tail(elem, acc) when is_list(elem) do
      ensure_evaled(elem, acc)
    end
    defp ensure_evaled_tail(elem, acc) do
      ensure_evaled_var(elem, acc)
    end

    defp ensure_evaled_var(elem, {i, ast}) do
      var = {String.to_atom("var" <> Integer.to_string(i)), [], __MODULE__}
      entry = quote do: (unquote(var) = unquote(elem))
      {var, {i + 1, [entry | ast]}}
    end
```

Note that the AST is kept as a list that needs to be reversed. Then you do:

```elixir
quote do
  unquote_splicing(:lists.reverse(ast))
  unquote(in_var(in_module?, left, &in_list(&1, h, t)))
end
```

This would also avoid introducing an empty list when all entries are literals.

I haven't tested it in any way and we need a lot of tests, especially for the tail handling ones.